substrate-node/pallets/pallet-smart-contract/src/migrations/v12.rs

```diff
@@ -0,0 +1,200 @@
+                )
+            );
+        }
+
+        check_contract_lock_v12::<T>()
+    }
+}
+
+#[cfg(feature = "try-runtime")]
+fn validate_pallet_version<T: Config>(
+    expected_version: types::StorageVersion,
+) -> Result<(), sp_runtime::TryRuntimeError> {
+    let current_version = PalletVersion::<T>::get();
+    debug!("current pallet version: {:?}", current_version);
+    ensure!(
+        current_version >= expected_version,
+        DispatchError::Other("Unexpected pallet version")
+    );
+    Ok(())
+}
+
+pub fn migrate_to_version_12<T: Config>() -> frame_support::weights::Weight {
+    debug!(
+        " >>> Starting contract pallet migration, pallet version: {:?}",
+        PalletVersion::<T>::get()
+    );
+
+    let mut total_reads = 0;
+    let mut total_writes = 0;
+
+    for (contract_id, contract) in Contracts::<T>::iter() {
+        log::debug!("Contract id: {:?}", contract_id);
+
+        if let Some(src_twin) = pallet_tfgrid::Twins::<T>::get(contract.twin_id) {
+            if ContractLock::<T>::contains_key(contract_id) {
+                let (r, w) = migrate_contract_lock::<T>(contract_id);
+                total_reads += r;
+                total_writes += w;
+
+                let (r, w) = remove_all_locks::<T>(&src_twin.account_id);
+                total_reads += r;
+                total_writes += w;
+            } else {
+                log::debug!("ContractLock not found for contract {:?}", contract_id);
+            }
+        } else {
+            log::debug!("Twin not found for contract {:?}", contract_id);
+        }
+    }
+
+    // Set the new storage version
+    PalletVersion::<T>::put(types::StorageVersion::V12);
+    total_writes += 1;
+
+    T::DbWeight::get().reads_writes(total_reads, total_writes)
+}
+
+fn migrate_contract_lock<T: Config>(contract_id: u64) -> (u64, u64) {
+    let mut reads = 0;
+    let mut writes = 0;
+
+    let old_contract_lock = ContractLock::<T>::take(contract_id);
+    reads += 1;
+    writes += 1;
+
+    ContractPaymentState::<T>::insert(
+        contract_id,
+        super::types::v12::ContractPaymentState {
+            standard_reserve: BalanceOf::<T>::zero(),
+            additional_reserve: BalanceOf::<T>::zero(),
+            standard_overdraft: old_contract_lock.amount_locked,
+            additional_overdraft: old_contract_lock.extra_amount_locked,
+            last_updated_seconds: old_contract_lock.lock_updated,
+            cycles: old_contract_lock.cycles,
+        },
+    );
+    writes += 1;
+
+    (reads, writes)
+}
+
+fn remove_all_locks<T: Config>(account_id: &T::AccountId) -> (u64, u64) {
+    let mut reads = 0;
+    let mut writes = 0;
+
+    let locks = pallet_balances::Pallet::<T>::locks(account_id);
+    reads += 1;
+
+    for lock in locks {
+        log::debug!("Removing lock: {:?} for account: {:?}", lock.id, account_id);
+        pallet_balances::Pallet::<T>::remove_lock(lock.id, account_id);
+        reads += 1;
+        writes += 1;
+    }
```

Comment:
Hmmm I just get a big doubt now!
What is unlocked here should not be reserved now ??

But I think I get my answer.
What was locked is transferred to the overdraft.
And the overdraft will be transferred to the reserve (if enough funds) at next billing cycle

Correct ?

Reply:
Exactly, we can't depend on previous state since we already know that it may inconsistent in some accounts.
Trying to check for inconsistency and handle it would be very expensive in the context of a storage migration.
The new logic would handle it progressively and accurately to ensure things is consistent again.